Rewrite x509.certificate_managed to be easier to use

glynnforrest opened the pull request:
### What does this PR do?
See https://github.com/saltstack/salt/issues/52167

After running into show-stopping bugs with `x509.certificate_managed` I decided to rewrite it. I have been running this function for the past 2 days with positive results.

### What issues does this PR fix or reference?

https://github.com/saltstack/salt/issues/52180, https://github.com/saltstack/salt/issues/39608, https://github.com/saltstack/salt/issues/41858 and others:

* Error messages from the `x509` module calls are written directly to the certificate file - fixed, the certificate file is only created when the `x509` module calls succeed.
* Certificates are created when no changes are required - fixed, the comparison logic has been updated.

**EDIT**: this does **not** attempt to correct https://github.com/saltstack/salt/issues/52167 - using a signing policy with `copypath` enabled. That is an issue with calling `x509.create_certificate` with `testrun` set to `True`. I will not be modifying the x509 *execution* module in this PR, just the *state module*.

Other improvements:

* Support for `test=True`
* Detailed messages explaining why the certificate needs to be changed (change in certificate information, cert has expired, not signed by the correct issuer, etc)
* The files specified in `append_certs` are checked to be valid certificates before appending them
* Better error messages when something goes wrong

I decided to remove the `managed_private_key` option after reading [this comment](https://github.com/saltstack/salt/issues/39608#issuecomment-342346894). Its removal significantly reduces the complexity of the function, and it can easily be replaced by an extra `x509.private_key_managed` call.

Screenshots! As well as showing the changes to the certificate information, it shows the reason the cert was updated.

![Screen Shot 2019-05-08 at 00 28 00](https://user-images.githubusercontent.com/578458/57339297-f08e5100-7128-11e9-9b16-2a5f16de1cff.png)
![Screen Shot 2019-05-08 at 00 28 41](https://user-images.githubusercontent.com/578458/57339298-f08e5100-7128-11e9-81c1-7b3e403df92c.png)
![Screen Shot 2019-05-08 at 00 29 34](https://user-images.githubusercontent.com/578458/57339299-f08e5100-7128-11e9-8951-d018c01821f3.png)

`test=True`:

![Screen Shot 2019-05-08 at 00 33 13](https://user-images.githubusercontent.com/578458/57339300-f126e780-7128-11e9-95ba-5d636cc82c1b.png)


### Tests written?
No

This is my first time writing code for salt and can't get the testsuite running (starting the salt-minion hangs). If I can get it working I'd be keen to add more tests covering this module.

### Commits signed with GPG?

Yes

glynnforrest commented:
One thing not fixed in this rewrite, the requirement of a dummy subjectAltName: https://github.com/saltstack/salt/issues/51869#issuecomment-472574844

That seems to be an issue with the x509 execution module.

Ch3LL commented:
@glynnforrest thanks for this :) . I'm more than willing to help you get the tests running. If you haven't seen it already we have some test documentation here: https://docs.saltstack.com/en/latest/topics/tutorials/writing_tests.html   Can you explain what you are attempting to run when you run into the issue when the salt-minion hangs? Also feel free to ping me in the community slack, my name is also ch3ll there and I can walk you through it. 

also @clinta can we get your review here?

alxwr commented:
@Ch3LL I too fail to run the test suite: https://github.com/saltstack/salt/pull/52456#issuecomment-498618310

Ch3LL commented:
@alxwr i responded in the other issue

clinta commented:
@Ch3LL My first look at this and it looks good.

I'll add another voice to the frustrations in running tests. I wanted to add a lot more x509 tests so that some of these bugs could be nailed down, but I've not been able to run them either. I think that posted documentation is out of date. See #51136.

I also agree with the idea of deprecating `managed_private_key`, that complicated things far more than was necessary.

Ch3LL commented:
@clinta thanks for the review and for bringing that issue to my attention. I went ahead and assigned myself, although I might not be able to get to it till next week. I do agree that we should definitely add some details about kitchen-salt as it makes things much easier, but a user can still run the test runner without using kitchen-salt. Can you elaborate on what your particular issues were when attempting  to run the tests? In the meantime I'll try to find some time next week to push out more details about kitchen-salt so its easier to setup the tests and run them.

glynnforrest commented:
Apologies for going off the grid. I'm still keen to get this tested well.

@Ch3LL My approach so far is to try and get `runtests.py --name=integration.states.test_x509` working. After installing the various requirements with pip I'm running into Python 2/3 issues. I haven't figured out a way to get m2crypto installed on python 3, and the test bootstrap doesn't seem to work on python 2.

**Run with python3** - test suite bootstraps OK, but both x509 tests are skipped because m2crypto isn't available. I can't figure out how to get it installed with pip3. It looks like python3 compatibility has been an issue in the past, here's an issue @thatch45 commented on: https://gitlab.com/m2crypto/m2crypto/issues/114

**Run with python2** - test suite has errors at the bootstrap stage:
```
Process MWorker-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/glynn/Desktop/salt/salt/utils/process.py", line 765, in _run
    return self._original_run()
  File "/home/glynn/Desktop/salt/salt/master.py", line 1141, in run
    self.__bind()
  File "/home/glynn/Desktop/salt/salt/master.py", line 1033, in __bind
    req_channel.post_fork(self._handle_payload, io_loop=self.io_loop)  # TODO: cleaner? Maybe lazily?
  File "/home/glynn/Desktop/salt/salt/transport/zeromq.py", line 704, in post_fork
    self.stream = zmq.eventloop.zmqstream.ZMQStream(self._socket, io_loop=self.io_loop)
  File "/usr/lib/python2.7/dist-packages/zmq/eventloop/zmqstream.py", line 114, in __init__
    self._init_io_state()
  File "/usr/lib/python2.7/dist-packages/zmq/eventloop/zmqstream.py", line 535, in _init_io_state
    self.io_loop.add_handler(self.socket, self._handle_events, self._state)
  File "/home/glynn/.local/lib/python2.7/site-packages/tornado/ioloop.py", line 910, in add_handler
    self._impl.register(fd, events | self.ERROR)
TypeError: argument must be an int, or have a fileno() method.
```
this repeats with various worker numbers (`Process MWorker-0`, `Process MWorker-1`, etc) until I kill the process.

Have you got a preferred way to update stale pull requests? Just merge develop?

@clinta I appreciate you taking a look over this. Glad to hear you support the removal of `managed_private_key`.

dmurphy18 commented:
@gyroplast Which version of m2crypto are you having issue with on Python 3 ?, have v0.31. and v0.33.on Python 3 running fine (RHEL 7).  What is the the OS the problem is occurring on ?

glynnforrest commented:
> Which version of m2crypto are you having issue with on Python 3 ?, have v0.31. and v0.33.on Python 3 running fine (RHEL 7). What is the the OS the problem is occurring on ?

@dmurphy18  I don't really understand python packaging, so hopefully doing something wrong that can be fixed easily.

Debian Stretch, `python3 --version` is `3.5.3`. I'm trying to install m2crypto with `pip3 install m2crypto`. At first it complained about missing openssl headers, so I installed Debian's `libssl-dev` package. Now it fails, something about 'swig'?

I'm happy to get a VM or vagrant box going if that'll make things easier.

dmurphy18 commented:
@glynnforrest I cannot find openssl/err.h on Debian 9 either, not even with backports enabled, however it is there on my Ubutnu 18.04.2  
root@david-lap2:/home/david# l /usr/include/openssl/err.h 
-rw-r--r-- 1 root root 11K Dec  5  2018 /usr/include/openssl/err.h

openssl:
  Installed: 1.1.0g-2ubuntu4.3
  Candidate: 1.1.0g-2ubuntu4.3
  Version table:
 *** 1.1.0g-2ubuntu4.3 500
        500 http://us.archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages
        500 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages
        100 /var/lib/dpkg/status
     1.1.0g-2ubuntu4 500
        500 http://us.archive.ubuntu.com/ubuntu bionic/main amd64 Packages

Sorry I can only suggest Ubuntu instead of Debian for this test at the moment. Will look further in the morning

OrangeDog commented:
With all the test work for 2019.2.1 it's probably worth trying to redo this on that branch? I'm also happy to test something manually.

glynnforrest commented:
OK, I'd be happy to rebase on the 2019.2.1 branch.

> With all the test work for 2019.2.1

Can you elaborate, are the tests going to be easier to run? Is there an issue I could take a look at? 

I did manage to get the suite running in an Ubuntu VM. Not giving up on this!

OrangeDog commented:
@glynnforrest 
I don't know any specifics, but that is the reason given for why it took six months to get the first bugfix release out.

Ch3LL commented:
@glynnforrest due to this new salt enhancement proposal https://github.com/saltstack/salt-enhancement-proposals/pull/20 we are merging everything into master. Do you mind migrating this PR over to the master branch? 

Also, we recently migrated to nox which should handle the requirements for you. 

To run the tests with nox, should be as easy as 

1. `pip install nox-py2` (even if your on py3)
2. `nox -e 'runtests-zeromq-2.7(coverage=False)' -- -n integration.states.test_x509`

glynnforrest commented:
> Do you mind migrating this PR over to the master branch?

Thanks @Ch3LL, I've been working on rebasing this branch onto master but it's a bit messy. There's some work on `develop` that's not on `master`, notably the `ext_mapping` option added by https://github.com/saltstack/salt/pull/49102:

* [develop](https://github.com/saltstack/salt/blob/develop/salt/states/x509.py#L338) (ext_mapping option included)
* [master](https://github.com/saltstack/salt/blob/master/salt/states/x509.py#L335) (ext_mapping option not included)

Looks like #49102 is ported to master in https://github.com/saltstack/salt/pull/54584. What would you like me to do?

glynnforrest committed:
Rewrite x509.certificate_managed to be easier to use

The function now displays clearer error messages when a problem occurs
and informative messages when comparing an existing certificate.

test=True is now supported.

It fixes #52180, #39608, #41858 and others:

* Error messages from the x509 module calls are written directly to
the certificate file - fixed, the certificate file is only created
when the x509 module calls succeed.
* Certificates are created when no changes are required - fixed, the
comparison logic has been updated.

The `managed_private_key` option has been removed due to the added
complexity. The functionality can easily be replicated with an
additional call to `x509.private_key_managed`. According to the comment
at https://github.com/saltstack/salt/issues/39608#issuecomment-342346894
`managed_private_key` has not worked since at least v2016.11.2.

Ch3LL commented:
thanks for doing that i really appreciate it. let me know when its ready for review or if you have any other test questions.

TOoSmOotH commented:
Any idea if this will make it in any time soon?

Ch3LL commented:
looks like we are just waiting on some tests getting written. anything i can help with @glynnforrest

glynnforrest committed:
Add integration test for a self-signed certificate

glynnforrest committed:
Test self signed certificates are renewed correctly

glynnforrest committed:
Clean up imports and debug logging

glynnforrest committed:
Use with_tempfile to prevent conflicts with other tests

glynnforrest committed:
Include more info about days_remaining and append_certs

glynnforrest commented:
> looks like we are just waiting on some tests getting written. anything i can help with @glynnforrest

Thanks for waiting, it's been hard to find the time to figure out how the tests work. I've written a couple of tests for a self signed cert, can you take a quick look @Ch3LL before I write any more? I'd like to test this pretty thoroughly but would like some initial feedback.

Ch3LL commented:
yeah let me give it a review

Ch3LL commented on salt/states/x509.py:
```diff
@@ -362,10 +366,112 @@ def csr_managed(name,
 
 def certificate_managed(name,
                         days_remaining=90,
-                        managed_private_key=None,
```
This would require users to update their states with no warning. Going forward we want to try to ensure we do not require users to update their state files, unless there is good reason. Is it possible to keep this same syntax for backward compatibility?

OrangeDog replied:
`managed_private_key` is currently broken and has not worked correctly for a while.

Ch3LL replied:
we still cannot rip out a kwarg entirely without warning, and I assume it was working at one point, so I still think we need to ensure we keep backwards compatibility.

glynnforrest replied:
It hasn't worked since at least v2016.11.2 according to https://github.com/saltstack/salt/issues/39608#issuecomment-342346894, I highly doubt anyone is actually using the kwarg successfully. I applaud the intent to keep backwards compatibility, but the only thing this rips out is a bug.

How about something like this:

```python
    if 'managed_private_key' in kwargs:
        raise salt.exceptions.SaltInvocationError(
            'managed_private_key is no longer supported by x509.certificate_managed, use a separate x509.private_key_managed call instead.')
```

Ch3LL replied:
can we not make a call to private_key_managed() if its set in kwargs? A user could be upgrading from v2016.11.2 so if we are going to rip it out I still think it needs to be properly deprecated: https://docs.saltstack.com/en/latest/topics/development/deprecations.html

OrangeDog replied:
If it was that simple then it wouldn't be broken.
I don't understand the rigid adherence to a deprecation schedule when it was already "ripped out" years ago. Even if you mark it deprecated now instead of removing it, it still won't work.

IMO it's bad design in the first place and should never have been added. Let each state manage its own thing, instead of also having one of them do both.

Ch3LL replied:
The rigid adherence is to ensure we do not rip out features from users and force them to update their states, unless there is good reason to do so.  I don't view a regression/bug as ripping out that feature, but something that should be fixed. From the comments I've read the reason to remove this feature is because it was not working since 2017.7 and to make the function more simple. I think the right approach would be to fix, but maybe i'm not understanding why its difficult to fix, since it was working at one time?

glynnforrest replied:
I appreciate the perspective from both of you. 

## Scenario

Let's actually consider what a user on 2016.11.2 will face when upgrading to 2017.7 or any other future version:

* Let's assume that `managed_private_key` did work in 2016.11.2, even though [this comment](https://github.com/saltstack/salt/issues/39608#issuecomment-342346894) doesn't actually guarantee that (_"at **least** v2016.11.2"_).
* Let's also assume that the only thing they have to worry about breaking is `x509.certificate_managed`, forgetting about the myriad of bugs and breaking changes that 2017.7 introduced that will **force the user to update their states**.

When upgrading to 2017.7 and using `x509.certificate_managed` with `managed_private_key`, the user will see an error. Nothing useful, just a stack trace dumped to the highstate output. There will be nothing in the changelog for 2017.7 telling them about this breaking change, no deprecation warnings will have been in place for the two versions before. They'll just have to search online, and probably stumble on one of the issues this PR is aiming to fix. **They will have to update their states**.

Same if they try and upgrade to any other version, including the current 3000.1. A Python stack trace with no obvious way to fix it.

## This PR improves things, it doesn't break anything

This pull request does not change any of that. `managed_private_key` is still broken in all previous versions since 2016.11, and none of the versions offer a way to fix it without the **user updating their states**.

What it does do is give a **useful error** instead of a stacktrace dumped to the highstate output. It tells the user upgrading from 2016.11 how to fix it (by updating their states) in a clear and efficient way. No searching online, no trawling through Github issues.

I'm going to guess that Salt's deprecation policy is for two reasons:

* To warn users about upcoming changes and help them to fix their states
* To reduce the support burden on the your team @Ch3LL.

In my view, showing a _clear_ and _actionable_ error message is 100% better than what is there currently.

> I think the right approach would be to fix

I'd agree if this problem had been identified in 2016.11 and fixed for 2017.7. At this point I think that ship has sailed. I can't imagine employees at SaltStack, nor people in the open-source community, will have the time to fix the bug for a version that is now beyond life support:

![Screen Shot 2020-04-08 at 19 29 11](https://user-images.githubusercontent.com/578458/78820401-a0221200-79cf-11ea-9500-4a9c954f24ce.png)

Since this PR is targeted against `master`, I believe this means it will go into a `3000.x` bugfix release, **not** 2017.7 or any other releases leading up to today. `managed_private_key` is irreversibly broken in those versions. If any users really need to be on those older versions with working x509, Salt can be patched quite easily to use this updated state (I myself am running this on 2019.2.3).

@Ch3LL I urge you to reconsider. I will not be adding `managed_private_key` back in my branch. You're welcome to use my code to patch it yourself, but I won't.

waynew replied:
Given that there's [apparently a workaround](https://github.com/saltstack/salt/issues/39608#issuecomment-285440745), that at *least* [two](https://github.com/saltstack/salt/issues/39608#issuecomment-285467363) [people](https://github.com/saltstack/salt/issues/39608#issuecomment-342346894) used with success, **I definitely agree with @Ch3LL here**.

It looks like it is possible to create a set of salt states with `managed_private_key` that works out of the box. Removing this without deprecation will break anyone who is currently using that workaround.

It would be different if there was no workaround and every single scenario involving `managed_private_key` failed to manage the key. Or am I misreading that issue?

glynnforrest replied:
> Given that there's apparently a workaround

@waynew, that is literally the exact workaround this PR is proposing in the error message.

OrangeDog replied:
@waynew You misunderstand. The state `certificate_managed` advertises a parameter (that doesn't work called `managed_private_key`). The workaround is to use the separate `private_key_managed` state instead, which does exactly what the broken parameter is supposed to do.

Removing the broken parameter will have no effect on anyone.

hatifnatt replied:
@waynew this workaround is working _by removing_ `managed_private_key` from `x509.certificate_managed` and creating separate state using `x509.private_key_managed`.

In my point of view there is a bug in current documentation, provided example with `managed_private_key` will not work. And this PR will fix this bug.

glynnforrest replied:
There seems to be some confusion here.

### Before

This is the usage we are removing. It doesn't work.

```yaml                                                         
/etc/pki/ca.crt:                                                
  x509.certificate_managed:                                     
    - signing_private_key: /etc/pki/ca.key                      
    - CN: ca.example.com                                                                             
    - basicConstraints: "critical CA:true"     
    - managed_private_key:
         name: /etc/pki/ca.key
         bits: 4096
         backup: True 
```

### After

This is the workaround and what this PR tells the user to do.

```yaml
/etc/pki/ca.key:                                                
  x509.private_key_managed:                                     
    - bits: 4096                                                
    - backup: True                                                                                     
                                                                
/etc/pki/ca.crt:                                                
  x509.certificate_managed:                                     
    - signing_private_key: /etc/pki/ca.key                      
    - CN: ca.example.com                                                                             
    - basicConstraints: "critical CA:true"                                                                 
    - require:                                                                                     
      - x509: /etc/pki/ca.key    
```

waynew replied:
This seems to work for me:

```
/tmp/sample.key.pem:
  x509.private_key_managed:
    - bits: 4096
    - backup: True


/tmp/sample.cert.pem:
  x509.certificate_managed:
    - signing_private_key: /tmp/sample.key.pem
    - CN: localhost
    - days_valid: 300
    - days_remaining: 300
    - managed_private_key:
        name: /tmp/sample.key.pem
        bits: 4096
```

Whether or not it's properly managing the private key, I'm not sure, but that state doesn't fail.

If we remove `managed_private_key` then it *will* fail, with no notice. That's a problem.

OrangeDog replied:
@waynew it's the `x509.private_key_managed` state that manages the key. Without it the `x509.certificate_managed` state will fail. The `managed_private_key` parameter isn't doing anything. Nobody would write what you've written except by accident. Even if you believed the documentation was correct you wouldn't write that.

If this change just logged a warning and ignored it instead of raising `SaltInvocationError` would that be OK?

glynnforrest replied:
Ah, good point @waynew, there are definitely examples (albeit mistaken) telling people to do that.

I'd be extremely happy to switch the `SaltInvocationError` to a logged warning, but getting the argument actually working will be a headache. Will that be OK?

waynew replied:
If we keep the existing behavior, and log a warning telling people that they probably shouldn't use the parameter, since it's useless, I'd be :+1: on that.

We can have a further discussion on whether we should fix or deprecate `managed_private_key` (if we have a different mechanism, it seems silly to have a second, broken way to do it).

glynnforrest replied:
Thanks, added in https://github.com/saltstack/salt/pull/52935/commits/f120c2257e6f9f1dbb58ea7f788cb5e7d41791e2 with a test case.

Ch3LL replied:
i'm good with this approach. :) Its a good medium of not breaking states and deprecating in the future. Thanks. Can I just ask that you add the deprecation to the `Aluminium` release so it has a warning in two major releases per the documentation here: https://docs.saltstack.com/en/latest/topics/development/deprecations.html

glynnforrest replied:
Sorry @Ch3LL I'm not 100% sure what you mean, I'm slow on the uptake 😄 

I read these pages:

https://docs.saltstack.com/en/latest/topics/development/deprecations.html
https://docs.saltstack.com/en/latest/topics/releases/version_numbers.html#code-names

>  For example, if the current release under development is Sodium, the deprecated code and associated warnings should remain in place and warn for at least Aluminum.

This is going in as a 3000.x (Neon) bugfix, right? Neon + 2 versions = Magnesium.

> Can I just ask that you add the deprecation to the Aluminium release

Wait, you don't want any warnings to show until then? `salt.utils.versions.warn_until` will log a warning until the specified version, then it will throw an error, not a deprecation.

Ch3LL replied:
No worries, i'm sure i was just not clear enough. This would be going into Sodium since its on the head of master, so Sodium + 2 versions would be Aluminium. We recently changed the branching and release strategy: https://github.com/saltstack/salt-enhancement-proposals/pull/20  The idea is we would do less point releases and release more often from master, a goal we are still trying to attain of every 4 months, and hopefully more often when more automation is put in place. A 3000.2 release would not occur unless there are critical enough issues that need to be fixed or CVE release.

I want the warning to show until the Aluminium release, instead of Magnesium. Hopefully that clears it up.

Ch3LL replied:
To clarify more, Im simply requesting Magnesium be changed to Aluminium instead.

glynnforrest replied:
Thanks for explaining!

> The idea is we would do less point releases and release more often from master, a goal we are still trying to attain of every 4 months, and hopefully more often when more automation is put in place. A 3000.2 release would not occur unless there are critical enough issues that need to be fixed or CVE release.

Oh great, I didn't realise there were no plans for another 3000.x release. Updated to Aluminium now.

OrangeDog commented on salt/states/x509.py:
```diff
@@ -362,10 +366,112 @@ def csr_managed(name,
+
+    try:
+        cert_info = __salt__['x509.read_certificate'](certificate=name)
+        required_cert_info = __salt__['x509.create_certificate'](testrun=True, **cert_spec)
```
It looks like this still creates new certificates before it knows whether they're needed (and then creates another one later the new one is needed)?
We had discussed this before in #52167.

glynnforrest replied:
I'm sorry @OrangeDog, I've misunderstood #52167 - I thought it was to do with the certificate comparisons (e.g. new cert created even though the requested `days_remaining` is `14` and the existing cert is `21`), not the `copypath` on the CA server filling up with certs.

I'm not going to attempt to fix #52167 here to keep the scope of this PR reasonable. Sorry! 

`x509.create_certificate` with `testrun=True` is designed to create a cert in memory that we can compare against here in the state. We should be able to call it with `testrun=True` as much as we need without filling up `copypath`. It's a bug in that execution module, which I'd like to fix in another PR if this one ever gets merged. I've updated the description of this PR to reflect this.

OrangeDog commented:
I don't recall what the problem was exactly right now, but in my custom replacement of this state I've removed the issuer checks - so it'll keep the old certificate if it's got a verifiable chain, regardless of what the chain actually is.

Perhaps that could/should be a new option in this rewrite?

glynnforrest committed:
Test self signed certificates properties are handled correctly

glynnforrest committed:
Explicitly remove support for managed_private_key

It doesn't work and hasn't for a long time, see
https://github.com/saltstack/salt/issues/39608#issuecomment-342346894.
Throwing a SaltInvocationError with a clear path to resolve the issue
is more useful.

glynnforrest committed:
Put test x509 state in the same directory as the others

glynnforrest commented:
From my perspective **this is ready**. @sagetherage this is rebased to be up to date with master. To address everyone's thoughts:

### This won't fix all x509 related bugs

This only addresses issues in the state module, not the execution module. In particular #52167 will not be fixed - I misunderstood the problem that bug was reporting.

The aim of this PR is to:

* Check certificate properties correctly
* Show better and more instructive error messages
* Make the code easier to read
* Be robust against errors coming from the x509 module
* Support `test=True`

, not to handle every x509 bug reported at present.

> I don't recall what the problem was exactly right now, but in my custom replacement of this state I've removed the issuer checks - so it'll keep the old certificate if it's got a verifiable chain, regardless of what the chain actually is.

> Perhaps that could/should be a new option in this rewrite?

@OrangeDog that's something I'd be open to doing in a future PR. I'm not looking to add new features at the moment and would like to get the state working as advertised first.

### About removing managed_private_key

It doesn't work and hasn't for a long time. Instead of throwing an unhelpful error, the state now throws a `SaltInvocationError` with a useful error message. @Ch3LL I really appreciate the aim to maintain backwards compatibility, but from what I can see this is actually an improvement for users. A state with `managed_private_key` will still throw an error, but a useful one instead.

### Tests

@Ch3LL I'm going to assume the tests I submitted are acceptable since you didn't comment on them. I've added a few more that tests the renewal logic of certs that have expired, changed subjectAltName, etc. I also did a bit of spring cleaning in `tests/integration/states/test_x509.py`.

### This state works and I'm ready to help

I've been using this rewrite for nearly a year to great success. I have a bit of time this week and would be happy to help get this one over the line! If anyone needs help running this locally for manual testing please let me know.

OrangeDog commented:
Just to note, the code I shared before fixes #52167 in the state; no changes to the execution module.

hatifnatt commented:
I'm setting up internal PKI using x509 state. During set up I faced with the problem described in #52167 - certificate is always issued and saved in `copypath` directory on a signing minion even with `test=True`.
@glynnforrest thanks for updated state, is seems like it fixing #52167 alt least new certificate not created in `copypath` directory on every state run. 
I will provide additional feedback after more use of updated state, if I find something worth mentioning. For now I can say that updated version is working and provide significant improvements.

glynnforrest committed:
Log a warning instead of throwing an error for managed_private_key

See https://github.com/saltstack/salt/pull/52935#discussion_r405839407

OrangeDog commented on salt/states/x509.py:
```diff
@@ -522,6 +526,10 @@ def certificate_managed(name, days_remaining=90, append_certs=None, **kwargs):
         They must be valid PEM files, otherwise an error will be thrown.
 
+    managed_private_key:
+        Has no effect and will be removed in Salt Magnesium.
```
Is it worth putting a "since v2016" to help avoid confusion about when it happened?

glynnforrest replied:
Nice idea, updated now 👍

Ch3LL commented:
This looks good to me. @waynew and @OrangeDog did you want to re-review here?

Also just a heads up, once this PR is merged https://github.com/saltstack/salt/pull/56578 we will need to rebase to ensure the new pre-commit tests pass. We recently added black and pre-commit to the code base. It looks like you might have already ran black against your code since i'm seeing double quotes. so thats great :)

glynnforrest commented:
> Also just a heads up, once this PR is merged #56578 we will need to rebase to ensure the new pre-commit tests pass. We recently added black and pre-commit to the code base. It looks like you might have already ran black against your code since i'm seeing double quotes. so thats great :)

Thanks for the heads up, I rebased onto latest master and ran black on every commit yesterday. Let me know if there's anything else to do after #56578. I tried installing the commit hook but it blew up for me at the `Lint Salt` stage (something about `AttributeError: module 'enum' has no attribute 'IntFlag'`). It did run black and isort fine however. Let me know if you'd like a stacktrace for the error.

Ch3LL commented:
thanks for doing that! for the lint failure, what version of python were you running so i can see if i can replicate it. And ya stacktrace would be great. We def want to make sure it works for everyone.

glynnforrest commented:
Rebased on top of #56578.

Here's the stacktrace @Ch3LL. It's on MacOS so I imagine that's why!

```
$ git commit
Linux Py2.7 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py2.7 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py2.7 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py2.7 Requirements.............................(no files to check)Skipped
Linux Py2.7 Crypto Requirements......................(no files to check)Skipped
Darwin Py2.7 Crypto Requirements.....................(no files to check)Skipped
Windows Py2.7 Crypto Requirements....................(no files to check)Skipped
Linux Py3.4 ZeroMQ Requirements......................(no files to check)Skipped
Cloud Py3.4 Requirements.............................(no files to check)Skipped
Linux Py3.4 Crypto Requirements......................(no files to check)Skipped
Linux Py3.5 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.5 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.5 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.5 Requirements.............................(no files to check)Skipped
Docs Py3.5 Requirements..............................(no files to check)Skipped
Linux Py3.5 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.5 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.5 Crypto Requirements....................(no files to check)Skipped
Lint Py3.5 Requirements..............................(no files to check)Skipped
Linux Py3.6 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.6 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.6 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.6 Requirements.............................(no files to check)Skipped
Docs Py3.6 Requirements..............................(no files to check)Skipped
Linux Py3.6 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.6 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.6 Crypto Requirements....................(no files to check)Skipped
Lint Py3.6 Requirements..............................(no files to check)Skipped
Linux Py3.7 ZeroMQ Requirements......................(no files to check)Skipped
Darwin Py3.7 ZeroMQ Requirements.....................(no files to check)Skipped
Windows Py3.7 ZeroMQ Requirements....................(no files to check)Skipped
Cloud Py3.7 Requirements.............................(no files to check)Skipped
Docs Py3.7 Requirements..............................(no files to check)Skipped
Linux Py3.7 Crypto Requirements......................(no files to check)Skipped
Darwin Py3.7 Crypto Requirements.....................(no files to check)Skipped
Windows Py3.7 Crypto Requirements....................(no files to check)Skipped
Lint Py3.7 Requirements..............................(no files to check)Skipped
isort....................................................................Passed
black....................................................................Passed
Lint Salt................................................................Failed
- hook id: nox-py2
- exit code: 1

Running session lint-salt-pre-commit
Processing pkg/osx/req.txt
Processing pkg/osx/req_ext.txt
Processing pkg/osx/req_pyobjc.txt
Processing requirements/static/darwin.in
pip install --progress-bar=off -r requirements/base.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/zeromq.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/pytest.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req_ext.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req_pyobjc.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/static/darwin.in --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/static/lint.in --constraint requirements/static/py3.7/lint.txt
Command pip install --progress-bar=off -r requirements/static/lint.in --constraint requirements/static/py3.7/lint.txt failed with exit code 1:
Collecting pylint==2.4.4 (from -c requirements/static/py3.7/lint.txt (line 13))
Downloading https://files.pythonhosted.org/packages/e9/59/43fc36c5ee316bb9aeb7cf5329cdbdca89e5749c34d5602753827c0aa2dc/pylint-2.4.4-py3-none-any.whl (302kB)
Collecting saltpylint==2019.11.14 (from -c requirements/static/py3.7/lint.txt (line 14))
Downloading https://files.pythonhosted.org/packages/fe/7f/0b5395b5d77f4d4157b327a42790179b1ec7bdfd7bd5dedbf2345b0cf6d6/SaltPyLint-2019.11.14-py2.py3-none-any.whl
Collecting mccabe==0.6.1 (from -c requirements/static/py3.7/lint.txt (line 10))
Using cached https://files.pythonhosted.org/packages/87/89/479dc97e18549e21354893e4ee4ef36db1d237534982482c3681ee6e7b57/mccabe-0.6.1-py2.py3-none-any.whl
Collecting astroid==2.3.3 (from -c requirements/static/py3.7/lint.txt (line 7))
Downloading https://files.pythonhosted.org/packages/ad/ae/86734823047962e7b8c8529186a1ac4a7ca19aaf1aa0c7713c022ef593fd/astroid-2.3.3-py3-none-any.whl (205kB)
Collecting isort==4.3.17 (from -c requirements/static/py3.7/lint.txt (line 8))
Downloading https://files.pythonhosted.org/packages/d3/57/2c79076db3c2aa9fc482628b8e70a9f6f2b4cb4b276636645b7080ac26ae/isort-4.3.17-py2.py3-none-any.whl (42kB)
Collecting modernize==0.5 (from -c requirements/static/py3.7/lint.txt (line 11))
Collecting pycodestyle==2.5.0 (from -c requirements/static/py3.7/lint.txt (line 12))
Using cached https://files.pythonhosted.org/packages/0e/0c/04a353e104d2f324f8ee5f4b32012618c1c86dd79e52a433b64fceed511b/pycodestyle-2.5.0-py2.py3-none-any.whl
Collecting lazy-object-proxy==1.4.3 (from -c requirements/static/py3.7/lint.txt (line 9))
Downloading https://files.pythonhosted.org/packages/07/3f/a3d687f83c7d44970f70ff0400677746c8860b11f0c08f6b4e07205f0cdc/lazy-object-proxy-1.4.3.tar.gz
Installing build dependencies: started
Installing build dependencies: finished with status 'error'
ERROR: Complete output from command /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/bin/python3.7 /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/_9/zj6803g95zqd5zpclzp823sm0000gn/T/pip-build-env-5ne1rp_z/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=30.3.0' wheel 'setuptools_scm>=3.3.1':
ERROR: Traceback (most recent call last):
File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"__main__", mod_spec)
File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip/__main__.py", line 16, in <module>
from pip._internal import main as _main  # isort:skip # noqa
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip/_internal/__init__.py", line 4, in <module>
import locale
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/locale.py", line 16, in <module>
import re
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/re.py", line 143, in <module>
class RegexFlag(enum.IntFlag):
AttributeError: module 'enum' has no attribute 'IntFlag'
----------------------------------------
ERROR: Command "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/bin/python3.7 /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/_9/zj6803g95zqd5zpclzp823sm0000gn/T/pip-build-env-5ne1rp_z/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=30.3.0' wheel 'setuptools_scm>=3.3.1'" failed with error code 1 in None
WARNING: You are using pip version 19.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Session lint-salt-pre-commit failed.

Lint Tests...............................................................Failed
- hook id: nox-py2
- exit code: 1

Running session lint-tests-pre-commit
Processing pkg/osx/req.txt
Processing pkg/osx/req_ext.txt
Processing pkg/osx/req_pyobjc.txt
Processing requirements/static/darwin.in
pip install --progress-bar=off -r requirements/base.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/zeromq.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/pytest.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req_ext.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r pkg/osx/req_pyobjc.txt --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/static/darwin.in --constraint requirements/static/py3.7/darwin.txt --constraint requirements/static/py3.7/darwin-crypto.txt
pip install --progress-bar=off -r requirements/static/lint.in --constraint requirements/static/py3.7/lint.txt
Command pip install --progress-bar=off -r requirements/static/lint.in --constraint requirements/static/py3.7/lint.txt failed with exit code 1:
Collecting pylint==2.4.4 (from -c requirements/static/py3.7/lint.txt (line 13))
Using cached https://files.pythonhosted.org/packages/e9/59/43fc36c5ee316bb9aeb7cf5329cdbdca89e5749c34d5602753827c0aa2dc/pylint-2.4.4-py3-none-any.whl
Collecting saltpylint==2019.11.14 (from -c requirements/static/py3.7/lint.txt (line 14))
Using cached https://files.pythonhosted.org/packages/fe/7f/0b5395b5d77f4d4157b327a42790179b1ec7bdfd7bd5dedbf2345b0cf6d6/SaltPyLint-2019.11.14-py2.py3-none-any.whl
Collecting isort==4.3.17 (from -c requirements/static/py3.7/lint.txt (line 8))
Using cached https://files.pythonhosted.org/packages/d3/57/2c79076db3c2aa9fc482628b8e70a9f6f2b4cb4b276636645b7080ac26ae/isort-4.3.17-py2.py3-none-any.whl
Collecting astroid==2.3.3 (from -c requirements/static/py3.7/lint.txt (line 7))
Using cached https://files.pythonhosted.org/packages/ad/ae/86734823047962e7b8c8529186a1ac4a7ca19aaf1aa0c7713c022ef593fd/astroid-2.3.3-py3-none-any.whl
Collecting mccabe==0.6.1 (from -c requirements/static/py3.7/lint.txt (line 10))
Using cached https://files.pythonhosted.org/packages/87/89/479dc97e18549e21354893e4ee4ef36db1d237534982482c3681ee6e7b57/mccabe-0.6.1-py2.py3-none-any.whl
Collecting modernize==0.5 (from -c requirements/static/py3.7/lint.txt (line 11))
Collecting pycodestyle==2.5.0 (from -c requirements/static/py3.7/lint.txt (line 12))
Using cached https://files.pythonhosted.org/packages/0e/0c/04a353e104d2f324f8ee5f4b32012618c1c86dd79e52a433b64fceed511b/pycodestyle-2.5.0-py2.py3-none-any.whl
Collecting lazy-object-proxy==1.4.3 (from -c requirements/static/py3.7/lint.txt (line 9))
Using cached https://files.pythonhosted.org/packages/07/3f/a3d687f83c7d44970f70ff0400677746c8860b11f0c08f6b4e07205f0cdc/lazy-object-proxy-1.4.3.tar.gz
Installing build dependencies: started
Installing build dependencies: finished with status 'error'
ERROR: Complete output from command /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/bin/python3.7 /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/_9/zj6803g95zqd5zpclzp823sm0000gn/T/pip-build-env-q2zb0ap_/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=30.3.0' wheel 'setuptools_scm>=3.3.1':
ERROR: Traceback (most recent call last):
File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
"__main__", mod_spec)
File "/usr/local/Cellar/python/3.7.4_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
exec(code, run_globals)
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip/__main__.py", line 16, in <module>
from pip._internal import main as _main  # isort:skip # noqa
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip/_internal/__init__.py", line 4, in <module>
import locale
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/locale.py", line 16, in <module>
import re
File "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/re.py", line 143, in <module>
class RegexFlag(enum.IntFlag):
AttributeError: module 'enum' has no attribute 'IntFlag'
----------------------------------------
ERROR: Command "/Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/bin/python3.7 /Users/glynn/.cache/pre-commit/repopr36g047/py_env-python3/lib/python3.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/_9/zj6803g95zqd5zpclzp823sm0000gn/T/pip-build-env-q2zb0ap_/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=30.3.0' wheel 'setuptools_scm>=3.3.1'" failed with error code 1 in None
WARNING: You are using pip version 19.1, however version 20.0.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Session lint-tests-pre-commit failed.
```

Ch3LL commented:
thanks for rebasing and posting that stacktrace. I created this issue https://github.com/saltstack/salt/issues/56603 so we can try to troubleshoot that there instead of adding noise to this PR.

Ch3LL commented:
ping @waynew can we get your re-review here and this will be ready to go

waynew reviewed (approved):
🚀 Thanks for working through that, looks great!

glynnforrest commented:
Thanks @waynew and @Ch3LL, great news! Want me to do anything with the outdated branch? Rebase or merge master?

waynew commented:
@glynnforrest Should be OK, I think we've been merging when there aren't any merge conflicts, just due to the length of time the test runs take.

We have some plans in place to reduce that here in the near future, but for now I don't think there's anything you need to do.

hatifnatt commented:
I found a problem with this updated state. It does not check is any `file.managed` options that are changed. I.e. if I change key / cert file properties like user, group, mode etc. no changes will be generated.

glynnforrest commented:
Thanks @hatifnatt, I'll get a fix for that written soon.

glynnforrest commented:
Can you try https://github.com/saltstack/salt/pull/56788 @hatifnatt?

hatifnatt commented:
@glynnforrest ok, I will try tomorrow.

hatifnatt commented:
I need to correct my previous statement, changing key file paramaters are correctly generate changes after this PR, in my case state was not executed because I have `prereq` requisite defined. Sorry for this.
But issue with certificate file parameters is still the case.
